astropy/time/formats.py

```diff
@@ -691,6 +691,156 @@ def to_value(self, timezone=None, parent=None):
+            # Code gets here for input as a dict or list of dict.  The dict input
+            # can be either scalar values or N-d arrays.  If the input was some
+            # other inappropriate object (giving dtype=object) then the first if-block
+            # will raise a ValueError.
+
+            # Turn it back into original object (a dict) or list of objects
+            val1_as_dict = val1.tolist()
+
+            if not isinstance(val1_as_dict, dict):
+                # Try converting list of dict to dict of list.  If this fails for any
+                # reason then the input does not meet API requirements for this format.
+                out = collections.defaultdict(list)
+                try:
+                    # If input is an iterable of dict then this will work, otherwise
+                    # the input was not compatible with YMDHMS.
+                    for row in val1_as_dict:
+                        for key, value in row.items():
+                            out[key].append(value)
+                except Exception:
+                    raise ValueError(f'input must be dict, list of dict, or table-like')
+
+                val1_as_dict = out
+
+            # Ensure the dict keys are good
+            if not set(val1_as_dict).issubset(set(ymdhms)):
+                raise ValueError(f'input key names {val1_as_dict.keys()}'
+                                 f'must be subset of {ymdhms}')
+
+            # Put this back into val1 and make all dict values into ndarray
+            for key, value in val1_as_dict.items():
+                val1_as_dict[key] = np.asarray(value)
+
+            # Check that all dict value shapes are the same
+            shapes = set([value.shape for value in val1_as_dict.values()])
+            if len(shapes) > 1:
+                raise ValueError(f'input values must all have the same shape, '
+                                 f'got {shapes}')
+
+        else:
+            raise ValueError('input must be dict, list of dict, or table-like')
+
+        return val1_as_dict, val2
+
+    def set_jds(self, val1, val2):
+        if val1 is None:
+            # Input was empty list []
+            jd1 = np.array([], dtype=np.float64)
+            jd2 = np.array([], dtype=np.float64)
+
+        else:
+            jd1, jd2 = erfa.dtf2d(self.scale.upper().encode('ascii'),
+                                  val1['year'],
+                                  val1.get('month', 1),
+                                  val1.get('day', 1),
+                                  val1.get('hour', 0),
+                                  val1.get('minute', 0),
+                                  val1.get('second', 0))
+
+        self.jd1, self.jd2 = day_frac(jd1, jd2)
+
+    @property
+    def value(self):
+        scale = self.scale.upper().encode('ascii')
+        iys, ims, ids, ihmsfs = erfa.d2dtf(scale, 9,
+                                           self.jd1, self.jd2_filled)
+        ihrs = ihmsfs['h']
```

Comment:
I think I'd use the parts of `ihmsfs` directly below, rather than define `ihrs`, etc.